### PR TITLE
Savanna

### PR DESCRIPTION
**Pull request by sirmax123**

### CODE REVIEW

File: deployment/puppet/savanna/manifests/dashboard.pp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+  } else {
+    $service_ensure = 'stopped'
+    $line_in_settings_py_ensure = 'absent'
+    $package_ensure = 'absent'
+  }
+
+
+
+  define line($file, $line, $ensure = 'present') {
```
Comment (aglarendil) [inline deployment/puppet/savanna/manifests/dashboard.pp]: This code is a little bit crappy. Look into puppet concat module or file_line resource to do such things. Or create separate python file and import it.

**Comment by dmitryilyin** [comment]
Yes, we all know that Puppet sucks with file editing and there are a lot of hacks to work around its architectural defects. But using sed patching is your last resort and should be avoided if there are better ways aviable.

If you just want to a line to file if this line is not present just use file_line resource from stdlib that is already aviable here.

Snipplets, templates and concats could be good for configuration files, but not for scripts usually.

There is also augeas that has it own... shortcomings.

You can also just create perl/ruby/python script if you need to perform complex modifications and run it with exec. But make shure that you make it idempotent.

This scripts could be implemented as native Puppet resources that could accept parameters and manage their state.

If all above fails or is too much for the job use sed/perl patching execs))

**Comment by dmitryilyin** [comment]
Now everything looks much better. Ready to be accepted.

File: deployment/puppet/savanna/manifests/dashboard.pp [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  file_line{'savannaurl'       : path => $local_settings, line => "${savanna_url_string}",                          ensure => $line_in_settings_py_ensure, notify => Service['httpd'], }
+
+  package { 'savanna_dashboard':
+    ensure => $package_ensure,
+    name   => $::savanna::params::savanna_dashboard_package_name,
+    notify => Service['httpd'],
+  }
+
+
+  service { 'httpd':
```
Comment (aglarendil) [inline deployment/puppet/savanna/manifests/dashboard.pp:44]: This service is duplicated.

**Pull request closed without merging**